# Sentiment Analysis Using Bag of Words and Decision Tree

## Project Overview

Natural Language Processing (NLP) enables machines to understand, analyze, and extract meaningful information from human language. One of the most fundamental NLP tasks is **Sentiment Analysis**, which aims to determine whether a piece of text expresses a positive or negative opinion.

In this project, I build a complete sentiment analysis pipeline using classical NLP techniques and Machine Learning algorithms. The workflow covers the entire process from raw text preprocessing to feature extraction, model training, evaluation, and experiment documentation.

The primary objective of this project is not only to achieve high predictive performance but also to systematically investigate how different preprocessing techniques, feature engineering strategies, and machine learning models affect sentiment classification performance.

---

## Dataset

Dataset Link: **[IMBD Movie Reviews](https://www.kaggle.com/datasets/mwallerphunware/imbd-movie-reviews-for-binary-sentiment-analysis)**

The dataset consists of movie reviews labeled with their corresponding sentiment:

* Positive
* Negative

The reviews contain real-world natural language, making them suitable for evaluating text preprocessing techniques and machine learning models.

---

## Project Pipeline

The following stages are implemented throughout this project:

### 1. Text Preprocessing

* Contraction expansion
* Lowercasing
* Text cleaning using regular expressions
* Stopword removal with preserved negations
* Lemmatization
* Corpus construction

### 2. Feature Extraction

* Bag of Words (CountVectorizer)
* N-gram generation
* Vocabulary analysis
* Feature selection using `max_features`
* Rare-word filtering using `min_df`

### 3. Model Training

Different machine learning algorithms are evaluated and compared, including:

* Gaussian Naive Bayes
* Multinomial Naive Bayes
* Bernoulli Naive Bayes
* Logistic Regression
* Support Vector Machines (SVM)
* Decision Trees
* Random Forests

### 4. Model Evaluation

Performance is assessed using multiple metrics:

* Accuracy
* Precision
* Recall
* F1-Score
* ROC-AUC Score
* Confusion Matrix
* Cross-Validation Mean Accuracy
* Cross-Validation Standard Deviation

---

## Experimental Approach

Rather than training a single model, this notebook follows an experimentation-driven methodology.

For each model, multiple configurations are tested, including different values for:

* `max_features`
* `min_df`
* N-gram ranges
* Preprocessing strategies

The goal is to identify the most effective configuration while understanding the trade-offs between model complexity, computational cost, and predictive performance.

---

## Key Learning Objectives

Through this project, I aim to:

* Develop a deeper understanding of NLP preprocessing techniques.
* Compare the behavior of different machine learning algorithms on text data.
* Analyze how feature engineering impacts classification performance.
* Build reproducible NLP pipelines suitable for real-world applications.
* Establish strong baselines before moving toward advanced embedding and transformer-based approaches.

---

**Author:** Hazem Mohamed

**Role:** AI Engineer | Machine Learning Engineer | NLP Engineer

**Repository:** [NLP Experimentation Lab](https://github.com/Hazem1695/NLP-Experimentation-Lab)


# **Importing the Libraries**

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# **Data preprocessing**

## Data Cleaning Check Template
This template is designed to quickly assess the quality of any dataset before building machine learning models or performing analysis.

It provides a structured overview of the dataset by checking for common data issues such as:

- Missing values

- Duplicate rows

- Incorrect data types

- Outliers

- Distribution of numerical features

- Categorical feature consistency

**What This Template Does**

- Displays basic dataset information (shape, data types)

- Identifies missing values and duplicates

- Summarizes numerical and categorical features

- Detects potential outliers using the IQR method

- Highlights columns with low unique values for quick inspection

How to Use

1. Load your dataset using Pandas  

2. Call the function:

In [2]:
def data_quality_report(df):

    print("DATA QUALITY REPORT")
    
    # Print a separator line for better readability
    
    print("=" * 50)
    print("BASIC INFO")
    print("=" * 50)
    
    # Show general information about the dataset (columns, data types, non-null values)
    print(df.info())
    
    # Show number of rows and columns
    print("\n" + "=" * 50)
    print("SHAPE OF DATA")
    print("=" * 50)
    print(f"Rows: {df.shape[0]}, Columns: {df.shape[1]}")
    
    # Check for missing (null) values in each column
    print("\n" + "=" * 50)
    print("MISSING VALUES")
    print("=" * 50)
    missing = df.isnull().sum()
    
    # Display only columns that have missing values
    print(missing[missing > 0])
    
    # Check for duplicate rows
    print("\n" + "=" * 50)
    print("DUPLICATES")
    print("=" * 50)
    print(f"Duplicate rows: {df.duplicated().sum()}")
    
    # Display data types of each column
    print("\n" + "=" * 50)
    print("DATA TYPES")
    print("=" * 50)
    print(df.dtypes)
    
    # Summary statistics for numerical columns (mean, std, min, max, etc.)
    print("\n" + "=" * 50)
    print("NUMERICAL SUMMARY")
    print("=" * 50)
    print(df.describe())
    
    # Summary for categorical (object) columns
    print("\n" + "=" * 50)
    print("CATEGORICAL SUMMARY")
    print("=" * 50)
    print(df.describe(include=['object']))
    
    # Show unique values for columns with low number of distinct values
    # Useful for detecting categories, errors, or inconsistencies
    print("\n" + "=" * 50)
    print("UNIQUE VALUES (LOW CARDINALITY)")
    print("=" * 50)
    for col in df.columns:
        if df[col].nunique() < 10:  # Only show columns with few unique values
            print(f"{col}: {df[col].unique()}")
            
    # correlation
    print("\n" + "=" * 50)
    print("CORRELATION MATRIX")
    print("=" * 50)
    print(df.corr(numeric_only=True))
    
    # Detect outliers using the IQR (Interquartile Range) method
    print("\n" + "=" * 50)
    print("OUTLIERS CHECK (IQR METHOD)")
    print("=" * 50)
    
    # Loop through only numerical columns
    for col in df.select_dtypes(include=np.number).columns:
        Q1 = df[col].quantile(0.25)  # 25th percentile
        Q3 = df[col].quantile(0.75)  # 75th percentile
        IQR = Q3 - Q1  # Interquartile range
        
        # Count rows that fall outside the normal range
        outliers = df[(df[col] < Q1 - 1.5 * IQR) | (df[col] > Q3 + 1.5 * IQR)]
        print(f"{col}: {len(outliers)} outliers")

## **Load dataset**
Apply Data Cleaning Check Template

In [ ]:
dataset = pd.read_csv('MovieReviewTrainingDatabase.csv')
data_quality_report(dataset)

DATA QUALITY REPORT
BASIC INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25000 entries, 0 to 24999
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   sentiment  25000 non-null  object
 1   review     25000 non-null  object
dtypes: object(2)
memory usage: 390.8+ KB
None

SHAPE OF DATA
Rows: 25000, Columns: 2

MISSING VALUES
Series([], dtype: int64)

DUPLICATES
Duplicate rows: 96

DATA TYPES
sentiment    object
review       object
dtype: object

NUMERICAL SUMMARY
       sentiment                                             review
count      25000                                              25000
unique         2                                              24904
top     Positive  You do realize that you've been watching the E...
freq       12500                                                  3

CATEGORICAL SUMMARY
       sentiment                                             review
count      25000                 

## Duplicate Data Detection

In [4]:
duplicates = dataset[dataset.duplicated(subset=['review'], keep=False)]
duplicates.sort_values('review')

,sentiment,review
21186,Negative,"Back in his youth, the old man had wanted to..."
21877,Negative,"Back in his youth, the old man had wanted to..."
14734,Negative,'Dead Letter Office' is a low-budget film abou...
5519,Negative,'Dead Letter Office' is a low-budget film abou...
7011,Positive,".......Playing Kaddiddlehopper, Col San Fernan..."
...,...,...
2685,Negative,"in this movie, joe pesci slams dunks a basketb..."
22244,Positive,it's amazing that so many people that i know h...
14767,Positive,it's amazing that so many people that i know h...
12462,Negative,this movie begins with an ordinary funeral... ...


## Quantifying Duplicate Review Frequencies

In [5]:
review_counts = dataset['review'].value_counts()
print("Reviews appearing more than once:")
print((review_counts > 1).sum())
print("\nMaximum repetitions:")
print(review_counts.max())

Reviews appearing more than once:
92

Maximum repetitions:
3


## Removing Duplicate Reviews & Resetting Index
> **Note:** This cell drops the repeated rows we identified in the previous steps and cleanly resets the row indices for model training

In [6]:
print("Before:", len(dataset))
dataset = dataset.drop_duplicates()
print("After:", len(dataset))
dataset = dataset.reset_index(drop=True)

Before: 25000
After: 24904


## Library Installation
> **Note:** The `contractions` library is required to automatically expand shortcuts like *don't* to *do not* and *I'm* to *I am* during preprocessing.

In [7]:
!pip install contractions

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 345.1/345.1 kB 6.5 MB/s eta 0:00:00:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 114.9/114.9 kB 7.4 MB/s eta 0:00:00


## **Cleaning the texts**

In [8]:
import nltk
nltk.download('stopwords')
nltk.download('wordnet')
nltk.download('omw-1.4')
import re
import contractions
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer

base_stopwords = set(stopwords.words('english'))
negation_words = {'not', 'no', 'never'}
all_stopwords = base_stopwords - negation_words

wnl = WordNetLemmatizer()

corpus = []

for i in range(0, len(dataset)):
    review = dataset['review'][i]
    # Fix contractions
    review = contractions.fix(review)
    # Lowercase
    review = review.lower()
    
    review = re.sub(r'[^a-zA-Z\s]', ' ', review)
    # Split
    words = review.split()
    # Chained Lemmatization (Handles both Verbs 'v' and Nouns 'n')
    review = [wnl.lemmatize(wnl.lemmatize(word, pos='v'), pos='n') for word in words if word not in all_stopwords]
    review = ' '.join(review) 
    corpus.append(review)

[nltk_data] Downloading package stopwords to /usr/share/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /usr/share/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /usr/share/nltk_data...


## Preprocessing Verification
> **Note:** Pulling the first two rows directly as a memory array to confirm that our lowercasing, stopword stripping, and lemmatization pipeline worked correctly before feeding it into the vectorizer.

In [9]:
# Pull the data directly as a fast memory array
raw_samples = dataset['review'].head(2).values

for i in range(2):
    print(f"=== REVIEW #{i+1} ===")
    print(f"RAW:     {raw_samples[i]}\n") 
    print(f"CLEANED: {corpus[i]}")
    print("-" * 50)

=== REVIEW #1 ===
RAW:     With all this stuff going down at the moment with MJ i've started listening to his music, watching the odd documentary here and there, watched The Wiz and watched Moonwalker again. Maybe i just want to get a certain insight into this guy who i thought was really cool in the eighties just to maybe make up my mind whether he is guilty or innocent. Moonwalker is part biography, part feature film which i remember going to see at the cinema when it was originally released. Some of it has subtle messages about MJ's feeling towards the press and also the obvious message of drugs are bad m'kay.  Visually impressive but of course this is all about Michael Jackson so unless you remotely like MJ in anyway then you are going to hate this and find it boring. Some may call MJ an egotist for consenting to the making of this movie BUT MJ and most of his fans would say that he made it for the fans which if true is really nice of him.  The actual feature film bit when it final

# **Creating the Bag of Word model**

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
# Add ngram_range=(1, 2) so it automatically catches phrases like "not good" or "no clue"
cv = CountVectorizer(max_features=10000, ngram_range=(1, 2))
X = cv.fit_transform(corpus).toarray()
y = dataset.iloc[:, 0].values

In [11]:
len(X[0])

10000

# **Encoding Categorical data Using Label Encoding**

In [12]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
y = le.fit_transform(y)

In [13]:
print(y)

[1 1 0 ... 0 0 1]


# Class Balance Check
> **Note:** Using NumPy to verify if our dataset is perfectly balanced between positive and negative reviews before splitting it into training and testing sets.

In [14]:
# This returns the unique classes and how many times they appear
classes, counts = np.unique(y, return_counts=True)
for c, count in zip(classes, counts):
    print(f"Class {c} contains {count}")

Class 0 contains 12432
Class 1 contains 12472


# **Splitting the dataset into the Training set and Test set**

In [15]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 0)

# Tuning Hyperparameters with RandomizedSearchCV

In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import RandomizedSearchCV
from sklearn.metrics import accuracy_score, classification_report

dt = DecisionTreeClassifier(random_state=0)

param_grid = {
    'criterion': ['gini', 'entropy', 'log_loss'],
    'splitter': ['best', 'random'],
    'max_depth': [None, 5, 10, 20, 30, 50],
    'min_samples_split': [2, 5, 10, 20],
    'min_samples_leaf': [1, 2, 4, 8],
    'max_features': [None, 'sqrt', 'log2']
}

In [ ]:
random_search = RandomizedSearchCV(
    estimator=dt,
    param_distributions=param_grid,
    n_iter=30,          # Number of random combinations
    scoring='accuracy',
    cv=5,
    random_state=42,
    n_jobs=-1,
    verbose=2
)

In [18]:
random_search.fit(X_train, y_train)

print(random_search.best_params_)
print(random_search.best_score_)

Fitting 5 folds for each of 30 candidates, totalling 150 fits
{'splitter': 'random', 'min_samples_split': 2, 'min_samples_leaf': 2, 'max_features': None, 'max_depth': 20, 'criterion': 'gini'}
0.7329718497578772
[CV] END criterion=entropy, max_depth=None, max_features=None, min_samples_leaf=4, min_samples_split=20, splitter=random; total time= 1.5min
[CV] END criterion=log_loss, max_depth=None, max_features=sqrt, min_samples_leaf=4, min_samples_split=2, splitter=random; total time=   3.9s
[CV] END criterion=entropy, max_depth=None, max_features=sqrt, min_samples_leaf=4, min_samples_split=10, splitter=best; total time=   2.8s
[CV] END criterion=entropy, max_depth=None, max_features=sqrt, min_samples_leaf=4, min_samples_split=10, splitter=best; total time=   2.8s
[CV] END criterion=log_loss, max_depth=20, max_features=sqrt, min_samples_leaf=8, min_samples_split=5, splitter=best; total time=   2.2s
[CV] END criterion=log_loss, max_depth=5, max_features=None, min_samples_leaf=2, min_samples

# **Training theDecision Tree model on the Training set**

In [19]:
from sklearn.tree import DecisionTreeClassifier
classifier = DecisionTreeClassifier(splitter='random', max_depth=20, min_samples_split=2, min_samples_leaf=2, criterion='gini', random_state=0)
classifier.fit(X_train, y_train)

DecisionTreeClassifier(max_depth=20, min_samples_leaf=2, random_state=0,
                       splitter='random')

# **Predicting the Test set results**

In [20]:
y_pred = classifier.predict(X_test)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[1 1]
 [1 1]
 [1 0]
 ...
 [0 0]
 [1 0]
 [1 0]]


# **Evaluating the Model Performance**

In [21]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report, roc_auc_score
from sklearn.model_selection import cross_val_score

print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

print("\nAccuracy Score:")
print(accuracy_score(y_test, y_pred))

print("\nClassification Report:")
print(classification_report(y_test, y_pred))

print("\nROC-AUC Score:")
print(roc_auc_score(y_test, y_pred))


accuracies = cross_val_score(estimator=classifier, X=X_train, y=y_train, cv=3)

print("\nMean Accuracy:")
print(accuracies.mean())

print("\nStandard Deviation:")
print(accuracies.std())

Confusion Matrix:
[[1700  820]
 [ 453 2008]]

Accuracy Score:
0.7444288295522987

Classification Report:
              precision    recall  f1-score   support

           0       0.79      0.67      0.73      2520
           1       0.71      0.82      0.76      2461

    accuracy                           0.74      4981
   macro avg       0.75      0.75      0.74      4981
weighted avg       0.75      0.74      0.74      4981


ROC-AUC Score:
0.7452658294795638

Mean Accuracy:
0.7327209757566631

Standard Deviation:
0.007165865685945954


# Decision Tree Performance Analysis for Text Classification

# 1. Objective

The objective of this experiment was to evaluate the effectiveness of the **Decision Tree** algorithm for binary text classification using **Bag-of-Words (BoW)** features generated by **CountVectorizer**.

Unlike linear models and probabilistic classifiers, Decision Trees partition the feature space through a sequence of hierarchical decisions. Although Decision Trees are highly interpretable, they are often challenged by high-dimensional and sparse text representations. This study investigates how feature extraction, tree complexity, and hyperparameter optimization influence their performance.

The experiment aims to answer the following research questions:

* How well does Decision Tree perform on sparse text data?
* Does increasing the vocabulary size improve performance?
* Can hyperparameter optimization significantly improve the classifier?
* Why was **RandomizedSearchCV** selected instead of **GridSearchCV**?
* What are the limitations of Decision Trees for text classification?

---

# 2. Experimental Setup

## Dataset

* Dataset cleaned before training.
* Duplicate documents removed.
* Final testing dataset: **4,981 documents**.

---

## Feature Extraction

Documents were transformed into numerical vectors using **CountVectorizer**.

The following feature extraction settings were evaluated:

| Configuration | Parameters                            |
| ------------- | ------------------------------------- |
| C1            | max_features=5000, ngram_range=(1,2)  |
| C2            | max_features=10000, ngram_range=(1,2) |
| C3            | max_features=15000, ngram_range=(1,2) |

The experiments progressively increased the vocabulary size to evaluate its impact on Decision Tree performance.

---

# 3. Hyperparameter Optimization Strategy

Initially, **GridSearchCV** was considered for hyperparameter optimization.

The complete search space contained:

* **1,728 parameter combinations**
* **5-fold Cross Validation**

Resulting in:

```text
1,728 × 5 = 8,640 model trainings
```

Training thousands of Decision Trees on high-dimensional text data would require a considerable amount of computation time.

Instead, **RandomizedSearchCV** was selected.

Configuration:

```text
30 random parameter combinations
5-fold Cross Validation
```

Resulting in only:

```text
30 × 5 = 150 model trainings
```

This reduced the computational cost by approximately **98%**, allowing efficient exploration of the hyperparameter space while maintaining competitive model performance.

---

# 4. Hyperparameters Explored

The optimization process searched over parameters including:

* Tree depth (`max_depth`)
* Splitting strategy (`splitter`)
* Splitting criterion (`gini`, `entropy`)
* Minimum samples required for splitting
* Minimum samples required per leaf
* Feature selection strategy

The best configurations discovered were:

### 5,000 Features

```python
DecisionTreeClassifier(
    splitter='random',
    criterion='gini',
    max_depth=30,
    min_samples_split=10,
    min_samples_leaf=8
)
```

---

### 10,000 Features

```python
DecisionTreeClassifier(
    splitter='random',
    criterion='gini',
    max_depth=20,
    min_samples_split=2,
    min_samples_leaf=2
)
```

---

### 15,000 Features

The same configuration remained optimal, indicating that the chosen tree structure generalized consistently as the vocabulary increased.

---

# 5. Experimental Results

| Configuration                   | Accuracy   | ROC-AUC    | CV Mean | CV Std |
| ------------------------------- | ---------- | ---------- | ------- | ------ |
| Manual Parameters (Entropy)     | 71.47%     | 0.7167     | 71.16%  | 0.0061 |
| RandomizedSearch (5K Features)  | 73.10%     | 0.7309     | 72.20%  | 0.0061 |
| RandomizedSearch (10K Features) | **74.44%** | **0.7453** | 73.27%  | 0.0072 |
| RandomizedSearch (15K Features) | 74.40%     | 0.7447     | 73.03%  | 0.0085 |
| Manual Tree (No Search)         | 71.87%     | 0.7186     | 71.53%  | 0.0015 |

---

# 6. Performance Analysis

## Effect of Hyperparameter Optimization

Hyperparameter tuning produced a clear improvement over manually selected parameters.

Accuracy increased from:

**71.47%**

to

**74.44%**

representing an improvement of nearly **3 percentage points**.

The optimized trees achieved better balance between underfitting and overfitting by selecting appropriate values for tree depth and minimum sample constraints.

---

## Why RandomizedSearchCV?

RandomizedSearchCV was selected because of its computational efficiency.

| Method             | Candidates | Total Fits |
| ------------------ | ---------: | ---------: |
| GridSearchCV       |      1,728 |      8,640 |
| RandomizedSearchCV |         30 |        150 |

The randomized search evaluated only **1.7%** of the total parameter combinations while discovering configurations that substantially improved model performance.

For high-dimensional NLP datasets, RandomizedSearchCV provides an excellent trade-off between search quality and computational cost.

---

## Effect of Vocabulary Size

Increasing the vocabulary size consistently improved performance.

| Features | Accuracy   |
| -------- | ---------- |
| 5K       | 73.10%     |
| **10K**  | **74.44%** |
| 15K      | 74.40%     |

The additional vocabulary provided more informative splitting features, enabling the tree to capture richer textual patterns.

However, the improvement between **10K** and **15K** features was minimal, suggesting that the model had already captured most of the discriminative information available to a single Decision Tree.

---

## Effect of Splitting Strategy

The optimized models consistently selected

```python
splitter='random'
```

rather than

```python
splitter='best'
```

Random splitting introduces additional randomness during tree construction.

This reduces overfitting and often produces better generalization on high-dimensional sparse datasets, where greedily selecting the locally optimal split may lead to overly specific decision boundaries.

---

# 7. Precision and Recall Analysis

For the best-performing model:

Class 0

* Precision = **0.79**
* Recall = **0.67**

Class 1

* Precision = **0.71**
* Recall = **0.82**

The classifier demonstrated noticeably higher recall for Class 1 than for Class 0.

This indicates that the Decision Tree identified most positive samples successfully but generated a relatively larger number of false positives for the negative class.

Although precision remained acceptable, the imbalance between recalls suggests that the tree's decision boundaries favored one class over the other.

---

# 8. ROC-AUC Analysis

The ROC-AUC score followed the same trend as classification accuracy.

| Configuration              | ROC-AUC    |
| -------------------------- | ---------- |
| Manual Parameters          | 0.7167     |
| RandomizedSearch (5K)      | 0.7309     |
| **RandomizedSearch (10K)** | **0.7453** |
| RandomizedSearch (15K)     | 0.7447     |

The optimized hyperparameters consistently improved the classifier's ability to separate the two target classes.

---

# 9. Cross-Validation Analysis

The best-performing model achieved:

* Mean Accuracy = **73.27%**
* Standard Deviation = **0.0072**

The relatively small standard deviation indicates that the optimized Decision Tree generalized reasonably well across different folds.

However, the overall predictive performance remained considerably lower than the strongest linear and probabilistic classifiers.

---

# 10. Best Model Configuration

The highest-performing Decision Tree was:

```python
CountVectorizer(
    max_features=10000,
    ngram_range=(1,2)
)

DecisionTreeClassifier(
    splitter='random',
    criterion='gini',
    max_depth=20,
    min_samples_split=2,
    min_samples_leaf=2,
    random_state=0
)
```

Performance:

* Accuracy = **74.44%**
* Precision = **0.75**
* Recall = **0.75**
* F1-score = **0.74**
* ROC-AUC = **0.7453**
* Cross-Validation Accuracy = **73.27%**
* Cross-Validation Standard Deviation = **0.0072**

---

# 11. Comparison with Previously Evaluated Models

| Model                   | Best Accuracy | ROC-AUC    |
| ----------------------- | ------------- | ---------- |
| Logistic Regression     | **88.92%**    | **0.8894** |
| LinearSVC               | 87.71%        | 0.8773     |
| Bernoulli Naive Bayes   | 87.66%        | 0.8767     |
| Multinomial Naive Bayes | 87.56%        | 0.8755     |
| Gaussian Naive Bayes    | 81.70%        | 0.8156     |
| K-Nearest Neighbors     | 75.31%        | 0.7523     |
| **Decision Tree**       | **74.44%**    | **0.7453** |

Decision Tree achieved the lowest performance among the optimized models evaluated in this study. While hyperparameter optimization improved its effectiveness, it remained substantially behind Logistic Regression, LinearSVC, and Naive Bayes.

---

# 12. Discussion

The experiments highlight the limitations of a single Decision Tree for sparse text classification. Decision Trees create axis-aligned splits based on individual features, whereas Bag-of-Words representations contain thousands of sparse dimensions with many zero values. As a result, the tree struggles to construct compact and generalizable decision boundaries.

Hyperparameter optimization using RandomizedSearchCV significantly improved performance while reducing computational cost compared with an exhaustive GridSearchCV. Increasing the vocabulary size also enhanced the quality of document representations, but gains became marginal beyond **10,000 features**.

These findings suggest that although Decision Trees are highly interpretable and easy to visualize, they are not well suited for high-dimensional NLP tasks. Ensemble methods such as Random Forest or Gradient Boosting typically overcome many of these limitations by combining the predictions of multiple trees.

---

# 13. Final Conclusion

This study evaluated the performance of Decision Trees for binary text classification using multiple Bag-of-Words feature configurations and extensive hyperparameter optimization.

The experiments demonstrated that **RandomizedSearchCV** provides a practical and computationally efficient alternative to GridSearchCV, reducing the number of required model evaluations from **8,640** to **150** while still identifying high-quality hyperparameter configurations.

The best-performing model employed **CountVectorizer with 10,000 features and bigrams**, together with an optimized **DecisionTreeClassifier** using random splitting and Gini impurity. This configuration achieved an **Accuracy of 74.44%**, an **ROC-AUC of 0.7453**, and a **Cross-Validation Accuracy of 73.27%**.

Although Decision Trees benefited from hyperparameter tuning and richer feature representations, their overall performance remained substantially below that of Logistic Regression, LinearSVC, and Naive Bayes. These results reinforce the conclusion that single Decision Trees are generally not the most suitable choice for high-dimensional sparse text classification, where linear and probabilistic models provide superior accuracy, robustness, and scalability.
